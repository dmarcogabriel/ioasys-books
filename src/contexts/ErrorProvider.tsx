import React, {createContext} from 'react';
import {UNAUTHORIZED_STATUS} from '../constants';
import Toast from 'react-native-toast-message';

type ErrorData = {
  message: string;
};

export interface ApiError {
  errors: ErrorData;
}

interface ErrorContextValue {
  handleError: (
    error: ApiError,
    status: number,
    logoutCallback: () => Promise<void>,
  ) => Promise<void>;
}

export const ErrorContext = createContext<ErrorContextValue>({
  handleError: async () => {},
});

interface ErrorProviderProps {
  children: React.ReactNode;
}

export const ErrorProvider = ({children}: ErrorProviderProps): JSX.Element => {
  const handleError = async (
    error: ApiError,
    status: number,
    logoutCallback: () => Promise<void>,
  ): Promise<void> => {
    let errorMessage: string = '';
    if (status === UNAUTHORIZED_STATUS) {
      await logoutCallback();
      errorMessage = 'Sess√£o expirada';
    } else {
      Toast.show({
        position: 'bottom',
        text1: errorMessage || error.errors.message,
        type: 'error',
      });
    }
  };

  return (
    <ErrorContext.Provider value={{handleError}}>
      {children}
    </ErrorContext.Provider>
  );
};
