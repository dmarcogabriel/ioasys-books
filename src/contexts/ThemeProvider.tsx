import React from 'react';
import {ThemeProvider as SCThemeProvider} from 'styled-components/native';

interface ThemeProviderProps {
  children: React.ReactNode;
}

export const ThemeProvider = ({children}: ThemeProviderProps): JSX.Element => {
  const theme = {
    colors: {
      primary: '#B22E6F',
      dark: '#333333',
      light: '#ffffff',
      grey: '#999999',
    },
  };

  return <SCThemeProvider theme={theme}>{children}</SCThemeProvider>;
};
