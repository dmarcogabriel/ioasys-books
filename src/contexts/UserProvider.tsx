import React, {createContext, useState, useEffect, useCallback} from 'react';
import {User} from '../@types/user.interface';
import {storeData, removeData, getData} from '../utils/storage';
import {AUTH_TOKEN_STORAGE_KEY} from '../constants';

interface UserContextValue {
  user: User | null;
  login: (user: User, token: string) => Promise<void>;
  logout: () => Promise<void>;
  token: string | null;
}

export const UserContext = createContext<UserContextValue>({
  user: null,
  token: null,
  login: async () => {},
  logout: async () => {},
});

interface UserProviderProps {
  children: React.ReactNode;
}

export const UserProvider = ({children}: UserProviderProps): JSX.Element => {
  const [user, setUser] = useState<User | null>(null);
  const [token, setToken] = useState<string | null>(null);
  const [isLoading, setIsLoading] = useState(true);

  const login = async (data: User, authToken: string): Promise<void> => {
    setUser(data);
    setToken(authToken);
    await storeData(AUTH_TOKEN_STORAGE_KEY, authToken);
  };

  const logout = async () => {
    await removeData(AUTH_TOKEN_STORAGE_KEY);
    setUser(null);
    setToken(null);
  };

  const loadToken = useCallback(async () => {
    setIsLoading(true);
    try {
      const authToken = await getData<string>(AUTH_TOKEN_STORAGE_KEY);
      if (authToken) {
        setToken(authToken);
      }
    } catch (error) {
      setIsLoading(false);
    }
    setIsLoading(false);
  }, []);

  useEffect(() => {
    loadToken();
  }, [loadToken]);

  return (
    <UserContext.Provider value={{user, token, login, logout}}>
      {!isLoading && children}
    </UserContext.Provider>
  );
};
