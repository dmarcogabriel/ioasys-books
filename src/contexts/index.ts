export * from './UserProvider';
export * from './ThemeProvider';
export * from './ErrorProvider';
