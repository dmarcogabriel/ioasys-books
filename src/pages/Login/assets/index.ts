import BackgroundImage from './background.png';

export default BackgroundImage;
