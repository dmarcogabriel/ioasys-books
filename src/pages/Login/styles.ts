import styled from 'styled-components/native';
import {ThemeProps} from '../../@types/theme.interface';

export const LoginContainer = styled.ImageBackground`
  height: 100%;
  justify-content: center;
  padding: 0 16px;
`;

export const LoginButton = styled.TouchableOpacity`
  background: ${({theme}: ThemeProps) => theme.colors.light};
  border-radius: 20px;
  padding: 8px 20px;
  align-items: center;
  justify-content: center;
`;

export const LoginButtonText = styled.Text`
  color: ${({theme}: ThemeProps) => theme.colors.primary};
  font-size: 16px;
`;
