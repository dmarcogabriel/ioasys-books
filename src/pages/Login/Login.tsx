import React from 'react';
import {useUser} from '../../hooks';
import {useFormik} from 'formik';
import {object, string} from 'yup';
import {HeaderTitle} from '../../components';
import BackgroundImage from './assets';
import {LoginContainer, LoginButton, LoginButtonText} from './styles';
import {TextInput} from './components';
import api from '../../services/api';
import {User} from '../../types/user.interface';
import Toast from 'react-native-toast-message';

export interface LoginInput {
  email: string;
  password: string;
}

export const Login = (): JSX.Element => {
  const {login} = useUser();

  const {values, handleChange, handleSubmit} = useFormik<LoginInput>({
    initialValues: {
      email: '',
      password: '',
    },
    validationSchema: object().shape({
      email: string().email().required(),
      password: string().required(),
    }),
    async onSubmit(formValues) {
      try {
        const {data, headers} = await api.post<User>(
          '/auth/sign-in',
          formValues,
        );
        await login(data, headers.authorization);
      } catch (error: any) {
        let errorMessage: string = '';
        if (error.response && error.response.data) {
          errorMessage = error.response.data.errors.message;
        }
        Toast.show({
          position: 'bottom',
          text1: errorMessage || 'Infelizmente, algo deu errado.',
          type: 'error',
        });
      }
    },
  });

  return (
    <LoginContainer source={BackgroundImage} resizeMode="cover">
      <HeaderTitle noPadding color="light" />
      <TextInput
        testID="emailInput"
        label="E-mail"
        value={values.email}
        onChange={handleChange('email')}
        returnKeyType="next"
      />
      <TextInput
        testID="passwordInput"
        label="Senha"
        value={values.password}
        onChange={handleChange('password')}
        secureTextEntry
        onEndEditing={handleSubmit}
        RightComponent={
          <LoginButton testID="loginButton" onPress={handleSubmit}>
            <LoginButtonText>Entrar</LoginButtonText>
          </LoginButton>
        }
      />
    </LoginContainer>
  );
};
