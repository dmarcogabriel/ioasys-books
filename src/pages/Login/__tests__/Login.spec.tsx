import React from 'react';
import {Login} from '../Login';
import {render, fireEvent, waitFor} from '@testing-library/react-native';
import {ThemeProvider} from '../../../contexts';
import renderer from 'react-test-renderer';

const mockLogin = jest.fn((data: any) => data);
jest.mock('../../../hooks', () => ({
  useUser: () => ({
    login: mockLogin,
  }),
}));

const LoginComponent = () => (
  <ThemeProvider>
    <Login />
  </ThemeProvider>
);

describe('pages/Login', () => {
  it('should fail login', async () => {
    const {getByTestId} = render(<LoginComponent />);

    const emailInput = getByTestId('emailInput');
    const passwordInput = getByTestId('passwordInput');
    const loginButton = getByTestId('loginButton');
    await waitFor(() => fireEvent.changeText(emailInput, 'wrong_email'));
    await waitFor(() => fireEvent.changeText(passwordInput, ''));
    await waitFor(() => fireEvent.press(loginButton));

    expect(mockLogin).not.toHaveBeenCalled();
  });

  it('should pass login', async () => {
    const {getByTestId} = render(<LoginComponent />);

    const emailInput = getByTestId('emailInput');
    const passwordInput = getByTestId('passwordInput');
    const loginButton = getByTestId('loginButton');
    await waitFor(() => fireEvent.changeText(emailInput, 'test@email.com'));
    await waitFor(() => fireEvent.changeText(passwordInput, '123456'));
    await waitFor(() => fireEvent.press(loginButton));

    expect(mockLogin).toHaveBeenCalled();
  });

  it('should match snapshot', () => {
    expect(renderer.create(<LoginComponent />).toJSON()).toMatchSnapshot();
  });
});
