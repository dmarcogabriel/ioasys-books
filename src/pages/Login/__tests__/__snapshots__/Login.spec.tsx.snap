// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Login should match snapshot 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Array [
      Object {
        "height": "100%",
        "justifyContent": "center",
        "paddingBottom": 0,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 0,
      },
    ]
  }
>
  <Image
    resizeMode="cover"
    source={
      Object {
        "testUri": "../../../src/pages/Login/assets/background.png",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": undefined,
          "width": undefined,
        },
        undefined,
      ]
    }
  />
  <View
    noMargin={false}
    noPadding={true}
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 48,
          "paddingTop": 0,
        },
      ]
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/components/HeaderTitle/assets/logo.png",
        }
      }
      style={
        Array [
          Object {},
        ]
      }
    />
    <Text
      color="light"
      style={
        Array [
          Object {
            "color": "#ffffff",
            "fontSize": 28,
            "marginLeft": 16,
          },
        ]
      }
    >
      Books
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#00000052",
          "borderRadius": 4,
          "marginBottom": 8,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 8,
          "paddingBottom": 8,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 8,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#ffffff",
              "opacity": 0.5,
            },
          ]
        }
      >
        E-mail
      </Text>
      <TextInput
        onChangeText={[Function]}
        returnKeyType="next"
        secureTextEntry={false}
        style={
          Object {
            "fontSize": 16,
          }
        }
        testID="emailInput"
        value=""
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#00000052",
          "borderRadius": 4,
          "marginBottom": 8,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 8,
          "paddingBottom": 8,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 8,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#ffffff",
              "opacity": 0.5,
            },
          ]
        }
      >
        Senha
      </Text>
      <TextInput
        onChangeText={[Function]}
        onEndEditing={[Function]}
        returnKeyType="default"
        secureTextEntry={true}
        style={
          Object {
            "fontSize": 16,
          }
        }
        testID="passwordInput"
        value=""
      />
    </View>
    <View
      style={
        Array [
          Object {
            "height": "100%",
            "justifyContent": "center",
            "paddingBottom": 16,
            "paddingLeft": 16,
            "paddingRight": 16,
            "paddingTop": 16,
            "position": "absolute",
            "right": 0,
          },
        ]
      }
      testID="textInputRightComponent"
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderRadius": 20,
            "justifyContent": "center",
            "opacity": 1,
            "paddingBottom": 8,
            "paddingLeft": 20,
            "paddingRight": 20,
            "paddingTop": 8,
          }
        }
        testID="loginButton"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#B22E6F",
                "fontSize": 16,
              },
            ]
          }
        >
          Entrar
        </Text>
      </View>
    </View>
  </View>
</View>
`;
